### DISCUSSION TITLE
Responsive Routes

### DISCUSSION THREAD

**Discussion by juangl** [remix-run/remix#1851] [Q&A] [answered]
What are the general thoughts about responsive routes in Remix? Right now I'm thinking about a use case that I have:

on desktop it renders:

<img width="500" src="https://user-images.githubusercontent.com/1887029/153100967-b3776d0c-b4d7-459f-9a96-20d95c4c34ad.png" />

on mobile it only renders route B:

<img width="200" src="https://user-images.githubusercontent.com/1887029/153101139-47c81f82-2ab4-46b1-8ee5-489d28a073fc.png" />

You usually use the user-gent on the server to predict what needs to be rendered (so that you don't over-fetch on mobile) but in Remix and the file system-based routing, this is not possible. Does the Remix team have any recommendations on how this should be implemented or there are future plans to have better support?

**Comment by sergiodxa** [comment]
You can make `/bills` render route A and `/bills/:id` render both, then with CSS you hide the Route A area on mobile devices

**Reply by juangl** [reply]
yeah, that's also the approach that I'm considering. I'm trying to figure out if there's a way to prevent e.g. the Route A data to be fetched on mobile.

**Comment by shamsup** [comment, marked as answer]
Don't use nested layouts for routes A and B, and only fetch what's needed in route B based on the user-agent.

```
routes/
  - bills.tsx // route A
  - bills.$id.tsx // route b
```

```ts
// routes/bills.$id
export const loader: LoaderFunction = ({request}) => {
  let userAgent = request.headers.get('user-agent');
  let aData = {}; // default empty state for route A data
  let bData = await loadRouteBData();
  if (!isMobile(userAgent)) {
    // get data for A portion of layout
    aData = await loadRouteAData();
  }
  return json({aData, bData});
}
```

You will have to duplicate the layout logic for the A route in both routes A and B, but it avoids the overfetching data issue. You would still hide the A layout portion of route B based on CSS media queries.

**Reply by shamsup** [reply]
It would be wise to have a post-hydration check for whether the data should actually be loaded though

**Reply by juangl** [reply]
I'll try this, thanks for the advice! I'd love to have a code-based routing configuration though.

**Reply by shamsup** [reply]
https://remix.run/docs/en/v1/api/conventions#routes
but I think this only runs at build time

**Reply by juangl** [reply]
awesome, will take a look. Thanks! :)
